### PR TITLE
Refactoring/orm

### PR DESCRIPTION
**Pull request by nibra**
# Reworking of the ORM

This PR provides a slightly different implementation of the ORM. The original version had the drawback, that it was not easily possible to inject different data sources (especially needed for testing).

Another important difference is, that entities can be plain objects. No interface needs to be implemented.
## RepositoryFactory

The programmatic entry-point to the ORM is the `RepositoryFactory`. It manages the creation of all necessary ORM internal dependancies. The factory is usually provided by the `StorageServiceProvider` with the key '`Repository`'.

The `RepositoryFactory` exposes one public method, `forEntity()`.

``` php
/** @var \Interop\Container\ContainerInterface $container */
$repository = $container->get('Repository')->forEntity($entityClassOrAlias);
```

You can inject a DataMapper, if you want to override the configuration settings.

``` php
/** @var \Interop\Container\ContainerInterface $container */
$repository = $container->get('Repository')->forEntity($entityClassOrAlias, $dataMapper);
```
## Repository

A `Repository` is created by the `RepositoryFactory`. It should not be instantiated directly outside the ORM, because that could do harm to the referential integrity.

``` php
/** @var \Interop\Container\ContainerInterface $container */
$repository = $container->get('Repository')->forEntity(Article::class);
```

The interface has changed a bit:

``` php
interface RepositoryInterface
{
    public function getById($id);
    public function findOne();
    public function findAll();
    public function add($entity);
    public function remove($entity);
    public function commit();
    public function getEntityClass()
    public function restrictTo($lValue, $op, $rValue);
}
```

`commit()` is a proxy to the `UnitOfWork`.
Since the repository is used as a collection in relations, the `restrictTo()` method is used preset conditions, so access is restricted to related entities.
## DataMapper

The `DataMapperInterface` is similar to the `RepositoryInterface`:

``` php
interface DataMapperInterface
{
    public function getById($id);
    public function findOne();
    public function findAll();
    public function insert($entity);
    public function update($entity);
    public function delete($entity);
}
```

Current implementations for this interface are `CsvDataMapper` and `DoctrineDataMapper`.
### CsvDataMapper

The `CsvDataMapper`'s constructor takes four arguments:
- the CsvDataGateway,
- the class of the entity,
- the basename of the data file, and
- the global entity registy.

``` php
$dataMapper = new CsvDataMapper(
    $gateway,
    Article::class,
    'articles',
    $entityRegistry
);
```
### DoctrineDataMapper

The `DoctrineDataMapper`'s constructor takes four arguments:
- the database connection,
- the class of the entity,
- the name of the table, and
- the global entity registy.

``` php
$dataMapper = new DoctrineDataMapper(
    $connection,
    Article::class,
    'articles',
    $entityRegistry
);
```
## EntityBuilder, EntityRegistry

The entities and their relations are managed by a couple of ORM internal classes. Userland code should never have to deal with them. 
### Configuration

The setup information for the ORM is currently stored in an `entities.ini` file.
## Testing

The tests have been re-organised, so they can be re-used for any storage type.
They are located in
- `tests/unit/ORM/StorageTestCases.php`,
  which is extended by `CsvStorageTest` and `DoctrineStorageTest`,
- `tests/unit/ORM/RelationTestCases.php`,
  extended by `CsvRelationTest` and `DoctrineRelationTest`, and
- `tests/unit/ORM/DataMapperTestCases.php`,
  extended by `CsvDataMapperTest` and `DoctrineDataMapperTest`.

This way, all tests are run for all data mappers, ensuring identical behaviour.
### Test Data

The test data is provided in `tests/unit/ORM/data/original`. To create an accessible copy for the tests for both CSV and SQLite, just run

``` bash
$ ./libraries/vendor/bin/robo create:testdata
```

When running tests using `robo`, the test data is (re)created automatically.

``` bash
$ ./libraries/vendor/bin/robo test
```

### CODE REVIEW

**Comment by laoneo** [comment]
Are you still working on this or can it be tested? Can it be tested with the front end as well or are unit tests the moment of truth? If the front end works, can you please add some instructions to the pr description how the set up can be done then. Thanks.

**Comment by nibra** [comment]
Currently, there's one issue left - handling relations that don't have an own ID. I'm working on it just now.

**Comment by laoneo** [comment]
This PR is quite big. Can it be tested with the front end as well or are unit tests the moment of truth? If the front end works, can you please add some instructions to the pr description how the set up can be done then and which url must be opened to view an article?

**Comment by nibra** [comment]
It is unit tests only for now. Frontend is WIP.

**Comment by laoneo** [comment]
But then I think it should not being merged as it is hard to get a feeling, how it will be used then. IMO it should be running at least in the current setup.

File: libraries/incubator/ORM/Service/RepositoryFactory.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	/**
+	 * RepositoryFactory constructor.
+	 *
+	 * @api
+	 *
+	 * @param   array                      $config     The configuration
+	 * @param   CsvDataGateway|Connection  $connection The connection / gateway
+	 * @param   TransactionInterface       $transactor A Transactor
+	 */
+	public function __construct(array $config, $connection, $transactor)
```
Comment (laoneo) [inline libraries/incubator/ORM/Service/RepositoryFactory.php:64]: Why is the connection passed as parameter, it is built, when the entity is loaded, or?

**Comment by laoneo** [comment]
I had a quick look on the files. I think the ORM part will work, but I can't see how it will fit into an existing application.

**Pull request merged**